### PR TITLE
Geometry_Engine: Performance tweaks and thread handling

### LINKED ISSUE #2797: Geometry_Engine: Stop checking for SelfIntersections when computing normal for curves?
#### Description:

While doing some profiling on some of our heavier geometry methods such as the Boolean intersection/difference/union methods for polycurves/polylines it was found that the clear bottleneck in the execution of those methods was coming from the computation of the normal of the curves, and more specifically, the check if the curves where self intersecting.

for this case, this check is fundamentally redundant, and the method itself simply just raises a warning if the curve actually is self intersecting, but is only raising a warning.

Commenting this check out and re-running saw a performance boost of 80% (!!) for the test case I was running, which involved quite a lot of heavy operations, including computation of voronoi diagrams and similar, with time to compute going from 10 to 2 seconds by just this simple change.

Given this, I am tempted to remove this check from the method, alternatively add another method (something like UnsafeNormal or similar) that simply skips this check (and maybe some others) that can be called when the self intersection check simply is not required for the method to perform the way that is needed.

Glad to get any input on this.

@al-fisher @pawelbaran @FraserGreenroyd

### LINKED ISSUE #2798: Geometry_Engine: Performance tweaks
#### Description:

Various small places have been identified as bottle necks in the code that can easily be refactored out to drastically reduce the cost of call to those methods without impacting functionality what so ever.

Those methods include
- RowEchelonForm (change cloning method)
- SplitAtPoints (pre-calculate some variables)
- LineIntersection (avoid cloning)
- BooleanIntersection (only check point in middle if required)

and some other lesser tweaks.

Might add more to the list and subsequent PR if found impactful

### LINKED ISSUE #2801: Geometry_Engine: RandomGeometry creates are not thread safe
#### Description:

The random geometry create methods are not thread safe as they all make use of the same instance of the Random class which is not itself thread safe. This can in some cases lead to weird crashes due to race conditions. The random create for vector is for example used in the IsContaining method which makes this important to fix to be able to parallelise any algorithm depending on that particular methods (which are tonnes, including boolean methods for polylines/polycurve)

#### Steps to reproduce:

Hard to reproduce with high reliability as with most race condition issues, but parallelising a method that relies on for example IsContaining leads to crashes sporadically happening.

#### Expected behaviour:

For thread saftey to be guarantied for the Create methods. A simple lock when extracting NextSeed should hopefully do.

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #2798 
Closes #2801 
Closes #2797 

Various performance tweaks speeding up various methods in the geometry engine with a significant amount for many cases.

Intention is for this PR to have not functional impact, but just getting rid of bottlenecks in the code by additional checks being made or code executed is slightly different order so that only calls required are actually made.

The methods touched in this PR are methods found to be bottlenecks for some particular heavy workflows, and changes in here have helped increase performance of those workflows by over 90% (sometimes quite a bit more than that).

Also attempting to make the Random geometry create methods at least more thread safe by protecting the random number generator behind a lock.

### Test files

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/BHoM_Engine/Geometry_Engine/%232802-PerformanceTweaks?csf=1&web=1&e=ZgjVcE

### CODE REVIEW

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance
@BHoMBot check installer

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check documentation-compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check documentation-compliance

File: Geometry_Engine/Query/LineIntersections.cs
Comment (pawelbaran) [inline Geometry_Engine/Query/LineIntersections.cs:311]: First of all, why is this method named `LineIntersections` if it is checking intersections between `Polylines`...?
```suggestion
        [Input("angleTolerance", "Angle tolerance to be used in the method. Used for checking if the polyline segments are colinear.", typeof(Angle))]
```

File: Geometry_Engine/Query/LineIntersections.cs
Comment (pawelbaran) [inline Geometry_Engine/Query/LineIntersections.cs:312]: Also this is not correct here

File: Geometry_Engine/Create/Arc.cs
Comment (pawelbaran) [inline Geometry_Engine/Create/Arc.cs:251]:
```suggestion
        private static object m_RandomLock = new object();
```

File: Geometry_Engine/Create/Plane.cs
Comment (pawelbaran) [inline Geometry_Engine/Create/Plane.cs:65]:
```suggestion
        [Description("Creates a random Plane based on a seed. If no seed is provided, a random one will be generated. If Box is provided, the resulting geometry will be contained within the box.")]
```

File: Geometry_Engine/Create/Plane.cs
Comment (pawelbaran) [inline Geometry_Engine/Create/Plane.cs:79]: You need to go ctrl+h _resuling_ => **_resulting_**

File: Geometry_Engine/Create/TransformMatrix.cs
Comment (pawelbaran) [inline Geometry_Engine/Create/TransformMatrix.cs:58]:
```suggestion
        [Input("translation", "The Vector corresponding to the seeked translation.")]
```

File: Geometry_Engine/Create/TransformMatrix.cs
Comment (pawelbaran) [inline Geometry_Engine/Create/TransformMatrix.cs:134]:
```suggestion
        [Description("Creates a TransformMatrix corresponding to a projection of the geometry on to the plane along a vector. If no vector is provided, the projection will be along the normal of the plane.")]
```

File: Geometry_Engine/Create/TransformMatrix.cs
Comment (pawelbaran) [inline Geometry_Engine/Create/TransformMatrix.cs:229]: Not sure if that makes sense, so please feel free to simply ignore (applies to all similar cases)
```suggestion
        [Input("minVal", "Minimum value to be used for each property in the created random matrix.")]
```

File: Geometry_Engine/Create/Vector.cs
Comment (pawelbaran) [inline Geometry_Engine/Create/Vector.cs:41]: (applies to all similar methods)
```suggestion
        [Output("vector", "The created Vector.")]
```

**Review by pawelbaran — CHANGES_REQUESTED** [review]
I finally made it! Went through the code, happy with the functional changes, but left a few comments on the descriptions... 🙈

File: Geometry_Engine/Compute/BooleanIntersection.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Compute/BooleanIntersection.cs:112]: THat would require checking analytics as well as message for deleted etc. All these descs already made this PR scope creep quite a bit. Prefer to not touch it.

File: Geometry_Engine/Compute/BooleanIntersection.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Compute/BooleanIntersection.cs:430]: This is not intended to have any functional change - The midpoint where compared on main as well (see first statement on main `mp.SquareDistance(rmp) <= sqTol` ). 

What this change is doing is to first check if the starts/ends are matching before the midpoint is extracted. Reason for this is that PointAtParameter is slower than start and end point, hence wanted to only extract it if necessary.

The way I read it (and intention) is main:

- Extract start, mid and end points
- Compare midpoints, if same, compare start and end points (allowing flips).

Branch:

- Extract start and end points
- Compare start and endpoints (allowing flips)
- If same extract mid point
- Compare midpoint

Could agree with you that we maybe should compare more points, but would prefer to do that on a separate PR, as the intention here is just to change the execution order, not the behaviour. If you still do not agree that the functionality should be the same, then I am happy to update (in case I have missed something 😄 )

File: Geometry_Engine/Compute/CullDuplicates.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Compute/CullDuplicates.cs:39]: Yes, good point, will add that to the description

File: Geometry_Engine/Create/Arc.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Create/Arc.cs:97]: no, should be less than 180 degrees. Not able to create arcs that are "larger" than a half circle basically, only arcs between a half-cricle and 0 angle

File: Geometry_Engine/Create/Arc.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Create/Arc.cs:196]: They are all public. I have used them quite a bit for testing purposes. Argument to be made to get them out of the core Geometry_Engine, and moved into the Test_Toolkit or similar, but would see that as a separate exercise.

Would say this PR is out of scope for this, even making them public (as that requires substantial versioning etc)

File: Geometry_Engine/Create/Arc.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Create/Arc.cs:238]: Will change to NextRandomSeed(). WOudl like to keep the "Next" word to align with the random convention. Hope you are ok with this?

File: Geometry_Engine/Create/BoundingBox.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Create/BoundingBox.cs:56]: Good point. Will raise. Just did not want to change the method functionality on this PR as I am not sure of the usage.

Issue raised here:

https://github.com/BHoM/BHoM_Engine/issues/2828

File: Geometry_Engine/Create/PolyCurve.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Create/PolyCurve.cs:52]: Dont think it can be actually. Each curve produced is using the endpoint of the previous as the start point

File: Geometry_Engine/Create/PolySurface.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Create/PolySurface.cs:52]: here I agree. WIll add

File: Geometry_Engine/Create/TransformMatrix.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Create/TransformMatrix.cs:229]: Will change to "element", using nomenclature from here https://en.wikipedia.org/wiki/Matrix_(mathematics)#Definition

File: Geometry_Engine/Create/Arc.cs
Comment (FraserGreenroyd) [reply Geometry_Engine/Create/Arc.cs:196]: I agree with moving them to Test_Toolkit as that is their primary purpose, but also agree that it's out of scope for this PR.

File: Geometry_Engine/Modify/SplitAtPoints.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Modify/SplitAtPoints.cs:45]: Tried to be more distinct where I could, but happy to change them all to the same if you think that is better?

File: Geometry_Engine/Query/IsContaining.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/IsContaining.cs:41]: Actually prefer to keep the current names. Something indicating that it is boxes you are working with on the side of the UI, and not like meshes/curves/breps.

If you really want this to be changed though, I can do it.

File: Geometry_Engine/Query/IsContaining.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/IsContaining.cs:72]: personally prefer the shorter names, but can be convinced otherwise.

Will fix only the full stop in this first runthrough

File: Geometry_Engine/Query/IsContaining.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/IsContaining.cs:108]: 😆

File: Geometry_Engine/Query/IsContaining.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/IsContaining.cs:136]: `Points not in the plane of the curve are deemed to be outside.` is trying to do that, but can rephrase if you think it is unclear?

File: Geometry_Engine/Query/IsContaining.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/IsContaining.cs:170]: helps the UI ensure that the parameter is the same as it used to be.

Given this only has one input oif this particular type, it might be able to fallback to find it anyway, but rather not take that risk :)

File: Geometry_Engine/Query/IsContaining.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/IsContaining.cs:231]: I think it only trigger now if the end and pPt are the same, making the direction into a 0 vector (given it is normalised)

Very likely this could be further refactored, but think this works as is now.

Happy to change if you disagree.

File: Geometry_Engine/Query/IsContaining.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/IsContaining.cs:439]: See above on boxes. Think it is useful to be able to spot the type you are working with.

File: Geometry_Engine/Query/IsEqual.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/IsEqual.cs:39]: Not sure here. @FraserGreenroyd ?

Comment (FraserGreenroyd) [reply Geometry_Engine/Query/IsEqual.cs:39]: I might argue a `within tolerance with each other.` However, I would also suggest that maybe the sentence as a whole be reworked.

`Checks if the two Planes are geometrically equal, i.e. if all of their properties match within a given tolerance.`

I don't think the `each other` bit is necessarily necessary owing to this method only checking two planes (based on the description, I have not read the code in full), so that it is doing it against each other is already a given I would think?

Comment (IsakNaslundBh) [reply Geometry_Engine/Query/IsEqual.cs:39]: Taken! 👍

File: Geometry_Engine/Query/LineIntersections.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/LineIntersections.cs]: Assume it is because it is doing linear intersections, but get what you mean.

WOuld prefer to not do any renaming in this PR though

File: Geometry_Engine/Create/Arc.cs
Comment (pawelbaran) [reply Geometry_Engine/Create/Arc.cs]: Me stupid, read it wrong, you are right.

File: Geometry_Engine/Create/Arc.cs
Comment (pawelbaran) [reply Geometry_Engine/Create/Arc.cs]: Sounds good 👍

File: Geometry_Engine/Modify/SplitAtPoints.cs
Comment (pawelbaran) [reply Geometry_Engine/Modify/SplitAtPoints.cs:45]: Personally I think `split` would be more informative as it indicated what has actually happened. But would not be nazi about it for sure 😉

File: Geometry_Engine/Compute/BooleanIntersection.cs
Comment (pawelbaran) [reply Geometry_Engine/Compute/BooleanIntersection.cs:112]: Ok, second thought on that: can imagine this being an optimisation for the case when one needs to intersect a single line with a collection of other lines: instead of doing the intersections and then summing them, one can do the union first and then intersect. Possibly more efficient, plus easier to call this method than call `foreach` every time one needs this feature.

File: Geometry_Engine/Compute/BooleanIntersection.cs
Comment (pawelbaran) [reply Geometry_Engine/Compute/BooleanIntersection.cs:430]: No no, you are right, I misread the code. No functional change is always welcome, particularly when the code works well (and I am pretty sure about that, remember @LMarkowski's struggle on that particular one 😃 😃 😃)

**Review by pawelbaran — DISMISSED** [review]
Happy to approve the PR now - one of my comments is not resolved yet, but it is not necessary to action it (in case @IsakNaslundBh does not come up with some ingenious naming system).

Just in case of me not being able to re-approve this after potential further tweaks: @FraserGreenroyd @IsakNaslundBh please be aware that IMO the PR is ready to merge, so feel free to merge it on that basis (no need for a full review from another person).

**Comment by pawelbaran** [comment]
@BHoMBot check required

File: Geometry_Engine/Modify/SplitAtPoints.cs
Comment (IsakNaslundBh) [reply Geometry_Engine/Modify/SplitAtPoints.cs]: Ok, thinking more about it, I can get behind that. Had to update some code for the null checks now anyway, so have pushed this as well :)

**Comment by IsakNaslundBh** [comment]
@BHoMBot check null-handling

**Comment by IsakNaslundBh** [comment]
@BHoMBot check null-handling

**Comment by IsakNaslundBh** [comment]
@BHoMBot check null-handling

**Comment by IsakNaslundBh** [comment]
@BHoMBot check null-handling

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check required

**Comment by IsakNaslundBh** [comment]
@BHoMBot check required

**Comment by IsakNaslundBh** [comment]
Will be seeking dispensation for the create method code compliance, as:

1. Completely out of scope of this PR
2. Debatable if changing the code will make things better. Have raised an issue https://github.com/BHoM/Test_Toolkit/issues/384 for this to be discussed how to best be handled going forward.

Discussed this with @FraserGreenroyd offline who agreed with this plan of action.

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am authorising dispensation to be granted on check ref. 6504963746

**Review by FraserGreenroyd — APPROVED** [review]
The only commits since [this review](https://github.com/BHoM/BHoM_Engine/pull/2802#pullrequestreview-977261606) have been compliance based. In agreement with @IsakNaslundBh and @pawelbaran I am approving this PR for merge and distribution owing to the testing conducted and reported on this PR by @pawelbaran .

**Comment by IsakNaslundBh** [comment]
@check ready-to-merge

**Comment by IsakNaslundBh** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**
